src/Stratis.Bitcoin.Features.Wallet/HdAddress.cs

```diff
@@ -24,15 +24,20 @@ public class TransactionCollection : ICollection<TransactionData>
 
         public AddressIdentifier GetAddressIdentifier()
         {
-            return this.repository.GetAddressIdentifier(this.wallet.Name, this.account.Name, this.address.AddressCollection.AddressType, this.address.Index);
+            return this.repository.GetAddressIdentifier(this.wallet.Name, this.account.Name, this.Address.AddressCollection.AddressType, this.Address.Index);
         }
 
         private IEnumerable<TransactionData> GetTransactions()
         {
-            if (this.repository == null)
-                return this.transactions;
+            // TODO: if (this.transactions == null)
```

Comment:
It seems like it would be a good idea to handle this case, if we don't then presumably we would lose the benefit of the transactions being cached when there are subsequent calls to `GetTransactions`?

Reply:
That being said, I suppose it is typically the `HdAddress` that will be getting used instead.

Reply:
> It seems like it would be a good idea to handle this case,

Unfortunately there is still a lot of code assuming that we are working with a "live" dataset instead of a disconnected dataset - I.e. code that expects that the details will change between calls. This may be best addressed in a separate PR.

Reply:
So I came across this issue looking at my bug on GetHistory and started to implement a fix. Would it not make more sense to add maybe a WalletId and maybe AccountId to HDPayment? We can then just load all of the payment data into a dictionary and do a fast lookup there when assigning the payment details.

Reply:
@ianadavies , you can add a method to the `HDPayment.cs` that joins the `HDPayment` table to the `HDTransactionTable` in a way that leverages the `UX_TransactionData_Output` unique index.

![image](https://user-images.githubusercontent.com/29645989/67026822-434a4b00-f154-11e9-88d6-ca843c32bd0b.png)